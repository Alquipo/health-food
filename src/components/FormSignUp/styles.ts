import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const FormWrapper = styled.div`
  ${({ theme }) => css`
    margin-top: ${theme.spacings.medium};
    width: 100%;
    display: flex;
    justify-content: center;

    form {
      width: 100%;

      display: grid;

      grid-template-columns: 1fr 1fr;
      grid-gap: ${theme.spacings.xxsmall};
      grid-template-areas: 'name birthDate' 'cpf zipCode' 'address number' 'district city' 'state state' 'button button';
      justify-items: center;

      margin: 0 ${theme.spacings.medium};

      ${media.greaterThan('large')`
        width: 50%;
      `};

      div:last-of-type {
        align-items: center;
      }

      ${media.lessThan('small')`
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;

      margin: 0 ${theme.spacings.medium};

      div {
        align-items: center;
      }
    `};
    }

    button {
      margin-top: ${theme.spacings.medium};
      margin-left: auto;
      grid-area: button;
      margin-bottom: ${theme.spacings.small};
    }

    label {
      margin-top: ${theme.spacings.xsmall};
    }
  `}
`
