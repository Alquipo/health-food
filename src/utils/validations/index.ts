import * as yup from 'yup'

const nascimentoRegex =
  /^((((19|[2-9]\d)\d{2})-(0[13578]|1[02])-(0[1-9]|[12]\d|3[01]))|(((19|[2-9]\d)\d{2})-(0[13456789]|1[012])-(0[1-9]|[12]\d|30))|(((19|[2-9]\d)\d{2})-02-(0[1-9]|1\d|2[0-8]))|(((1[6-9]|[2-9]\d)(0[48]|[2468][048]|[13579][26])|((16|[2468][048]|[3579][26])00))-02-29))$/g

const cpfRegex = /^\d{3}.?\d{3}.?\d{3}-?\d{2}$/

const cepRegex = /^([\d]{2})\.*([\d]{3})-*([\d]{3})/

const estadoRegex =
  /^(AC|AL|AP|AM|BA|CE|DF|ES|GO|MA|MT|MS|MG|PA|PB|PR|PE|PI|RJ|RN|RS|RO|RR|SC|SP|SE|TO)$/i

export const fieldsValidationSchema = yup.object({
  name: yup.string().required('Is not allowed to be empty'),
  birthDate: yup
    .string()
    .required('Is not allowed to be empty')
    .matches(nascimentoRegex, 'must be a valid email'),
  cpf: yup
    .string()
    .max(14, 'Must be 14 characters')
    .required('Is not allowed to be empty')
    .matches(cpfRegex, 'Invalid CPF'),
  zipCode: yup
    .string()
    .required('Is not allowed to be empty')
    .matches(cepRegex, 'Invalid ZipCode'),
  address: yup.string().required('Is not allowed to be empty'),
  district: yup.string().required('Is not allowed to be empty'),
  number: yup.number().required('Is not allowed to be empty'),
  city: yup.string().required('Is not allowed to be empty'),
  state: yup
    .string()
    .required('Is not allowed to be empty')
    .matches(estadoRegex, 'Invalid State')
})
