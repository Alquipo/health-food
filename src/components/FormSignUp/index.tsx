import * as S from './styles'

import { useRouter } from 'next/router'

import { Form, Formik, Field } from 'formik'

import { fieldsValidationSchema } from '../../utils/validations'

import Button from '../../components/Button'
import TextField from '../../components/TextField'
import useStorage from '../../hooks/use-storage'

export type FormType = {
  name: string
  birthDate: string
  cpf: string
  zipCode: string
  address: string
  district: string
  number: string
  state: string
  city: string
}

const initialValues: FormType = {
  name: '',
  birthDate: '',
  cpf: '',
  zipCode: '',
  address: '',
  district: '',
  number: '',
  state: '',
  city: ''
}

function FormSignUp() {
  const { setLocalStorage, setCookie } = useStorage()
  const { push } = useRouter()

  /* istanbul ignore next */
  async function handleSubmit(values: FormType) {
    setLocalStorage('signup-data', values)
    setCookie('signup-data', values, 5)

    window.alert('registered successfully')
    push(`/`)
  }

  /* istanbul ignore next */
  async function fetchCEP(
    e: React.FocusEvent<HTMLInputElement>,
    setFieldValue: (field: string, value: string) => void
  ) {
    const { value } = e.target as HTMLInputElement

    const regex = new RegExp(/[^0-9]|[/ /]/g, '')
    const sCep = value.trim().replace(regex, '')

    if (sCep.length !== 8) return

    const resp = await fetch(`https://viacep.com.br/ws/${value}/json`, {
      method: 'GET',
      headers: { 'content-type': 'application/json' }
    })

    const data = await resp.json()

    setFieldValue('address', data.logradouro)
    setFieldValue('district', data.bairro)
    setFieldValue('city', data.localidade)
    setFieldValue('state', data.uf)
  }

  return (
    <>
      <S.FormWrapper>
        <Formik
          initialValues={initialValues}
          onSubmit={handleSubmit}
          validationSchema={fieldsValidationSchema}
          validateOnBlur={false}
          validateOnChange={false}
        >
          {({ setFieldValue }) => (
            <Form autoComplete="off">
              <Field as={TextField} name="name" label="Name" type="text" />

              <Field
                as={TextField}
                name="birthDate"
                label="Birth Date"
                type="date"
              />

              <Field
                as={TextField}
                mask="999.999.999-99"
                name="cpf"
                label="CPF"
                type="text"
              />

              <Field
                as={TextField}
                name="zipCode"
                label="Zip Code"
                mask="99999-999"
                type="text"
                onBlur={(e: React.FocusEvent<HTMLInputElement>) =>
                  fetchCEP(e, setFieldValue)
                }
              />

              <Field
                as={TextField}
                name="address"
                label="Address"
                type="text"
              />

              <Field
                as={TextField}
                name="number"
                label="Number"
                type="number"
              />

              <Field
                as={TextField}
                name="district"
                label="District"
                type="text"
              />

              <Field as={TextField} name="city" label="City" type="text" />

              <Field
                as={TextField}
                name="state"
                label="State (UF)"
                type="text"
              />

              <Button fullSize type="submit" aria-label="Confirm">
                Confirm
              </Button>
            </Form>
          )}
        </Formik>
      </S.FormWrapper>
    </>
  )
}

export default FormSignUp
