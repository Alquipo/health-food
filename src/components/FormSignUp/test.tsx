import { fireEvent, render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import FormSignUp from '.'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter')
const push = jest.fn()

useRouter.mockImplementation(() => ({
  push,
  query: '',
  asPath: '',
  route: '/'
}))

describe('<FormSignUp />', () => {
  beforeEach(() => {
    window.localStorage.clear()
  })
  it('should render the form', () => {
    const { container } = render(<FormSignUp />)

    expect(screen.getByTestId(/name/i)).toBeInTheDocument()
    expect(screen.getByTestId(/birthDate/i)).toBeInTheDocument()
    expect(screen.getByTestId(/CPF/i)).toBeInTheDocument()
    expect(screen.getByTestId(/ZipCode/i)).toBeInTheDocument()
    expect(screen.getByTestId(/Address/i)).toBeInTheDocument()
    expect(screen.getByTestId(/number/i)).toBeInTheDocument()
    expect(screen.getByTestId(/District/i)).toBeInTheDocument()
    expect(screen.getByTestId(/City/i)).toBeInTheDocument()
    expect(screen.getByTestId(/State/i)).toBeInTheDocument()

    expect(screen.getByRole('button', { name: /Confirm/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should be accessible with tab', () => {
    render(<FormSignUp />)

    expect(document.body).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/name/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/birthDate/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/CPF/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/ZipCode/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/Address/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/number/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/District/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/City/i)).toHaveFocus()

    userEvent.tab()

    expect(screen.getByTestId(/State/i)).toHaveFocus()
  })

  it('should handle submit button ', () => {
    render(<FormSignUp />)

    const handleSubmit = jest.fn()

    fireEvent.click(screen.getByLabelText(/Confirm/i))

    waitFor(() => {
      expect(handleSubmit).toHaveBeenCalled()
      jest.spyOn(window, 'alert')
    })
  })
})
