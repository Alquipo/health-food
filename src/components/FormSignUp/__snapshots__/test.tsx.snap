// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignUp /> should render the form 1`] = `
.c0 {
  margin-top: 3.2rem;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 form {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 0.8rem;
  grid-template-areas: 'name birthDate' 'cpf zipCode' 'address number' 'district city' 'state state' 'button button';
  justify-items: center;
  margin: 0 3.2rem;
}

.c0 form div:last-of-type {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 button {
  margin-top: 3.2rem;
  margin-left: auto;
  grid-area: button;
  margin-bottom: 2.4rem;
}

.c0 label {
  margin-top: 1.6rem;
}

.c12 {
  padding: 1.6rem 2.4rem;
  background: #BADC58;
  color: #FFFFFF;
  border-radius: 0.6rem;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
  cursor: pointer;
  border: none;
  width: 100%;
  font-size: 1.8rem;
  font-weight: 600;
}

.c12:hover {
  -webkit-filter: brightness(0.95);
  filter: brightness(0.95);
}

.c1 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: name;
}

.c4 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: birthDate;
}

.c5 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: cpf;
}

.c6 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: zipCode;
}

.c7 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: address;
}

.c8 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: number;
}

.c9 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: district;
}

.c10 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: city;
}

.c11 {
  width: 100%;
  max-width: 25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  grid-area: state;
}

.c2 {
  display: block;
  font-size: 1.6rem;
  margin-bottom: 0.8rem;
}

.c3 {
  display: block;
  width: 100%;
  padding: 1.6rem;
  border: solid 0.1rem #1D164D;
  border-radius: 0.6rem;
  font-size: 1.6rem;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
  font-size: 1.4rem;
}

.c3:focus,
.c3:hover {
  outline: none;
}

@media (min-width:1170px) {
  .c0 form {
    width: 50%;
  }
}

@media (max-width:450px) {
  .c0 form {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin: 0 3.2rem;
  }

  .c0 form div {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div
  class="c0"
>
  <form
    action="#"
    autocomplete="off"
  >
    <div
      class="c1"
    >
      <label
        aria-labelledby="Name"
        class="c2"
        for="name"
      >
        Name
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="name"
        name="name"
        type="text"
        value=""
      />
    </div>
    <div
      class="c4"
    >
      <label
        aria-labelledby="Birth Date"
        class="c2"
        for="birthDate"
      >
        Birth Date
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="birthDate"
        name="birthDate"
        type="date"
        value=""
      />
    </div>
    <div
      class="c5"
    >
      <label
        aria-labelledby="CPF"
        class="c2"
        for="cpf"
      >
        CPF
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="cpf"
        name="cpf"
        type="text"
        value=""
      />
    </div>
    <div
      class="c6"
    >
      <label
        aria-labelledby="Zip Code"
        class="c2"
        for="zipCode"
      >
        Zip Code
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="zipCode"
        name="zipCode"
        type="text"
        value=""
      />
    </div>
    <div
      class="c7"
    >
      <label
        aria-labelledby="Address"
        class="c2"
        for="address"
      >
        Address
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="address"
        name="address"
        type="text"
        value=""
      />
    </div>
    <div
      class="c8"
    >
      <label
        aria-labelledby="Number"
        class="c2"
        for="number"
      >
        Number
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="number"
        name="number"
        type="number"
        value=""
      />
    </div>
    <div
      class="c9"
    >
      <label
        aria-labelledby="District"
        class="c2"
        for="district"
      >
        District
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="district"
        name="district"
        type="text"
        value=""
      />
    </div>
    <div
      class="c10"
    >
      <label
        aria-labelledby="City"
        class="c2"
        for="city"
      >
        City
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="city"
        name="city"
        type="text"
        value=""
      />
    </div>
    <div
      class="c11"
    >
      <label
        aria-labelledby="State (UF)"
        class="c2"
        for="state"
      >
        State (UF)
      </label>
      <input
        autocomplete="off"
        class="c3"
        data-testid="state"
        name="state"
        type="text"
        value=""
      />
    </div>
    <button
      aria-label="Confirm"
      class="c12"
      type="submit"
    >
      Confirm
    </button>
  </form>
</div>
`;
